crates/ruma-events/src/room/power_levels.rs

```diff
@@ -298,57 +316,255 @@ impl RedactedStateEventContent for RedactedRoomPowerLevelsEventContent {
-/// power levels of a room, regardless of whether the most recent power-levels event is redacted or
-/// not.
+/// This struct contains all the power levels settings from the specification and can be constructed
+/// from several [`RoomPowerLevelsSource`]s, which means that it can be used when wanting to inspect
+/// the power levels of a room, regardless of whether the most recent power levels event is redacted
+/// or not, or the room has no power levels event.
+///
+/// This can also be used to change the power levels of a room by mutating it and then converting it
+/// to a [`RoomPowerLevelsEventContent`] using `RoomPowerLevelsEventContent::try_from` /
+/// `.try_into()`. This allows to validate the format of the power levels before sending them. Note
+/// that the homeserver might still refuse the power levels changes depending on the current power
+/// level of the sender.
 #[derive(Clone, Debug)]
 #[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
 pub struct RoomPowerLevels {
     /// The level required to ban a user.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to `50`.
     pub ban: Int,
 
     /// The level required to send specific event types.
     ///
     /// This is a mapping from event type to power level required.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to an empty map.
     pub events: BTreeMap<TimelineEventType, Int>,
 
     /// The default level required to send message events.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to `0`.
     pub events_default: Int,
 
     /// The level required to invite a user.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to `0`.
     pub invite: Int,
 
     /// The level required to kick a user.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to `50`.
     pub kick: Int,
 
     /// The level required to redact an event.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to `50`.
     pub redact: Int,
 
     /// The default level required to send state events.
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`], defaults to `50`.
     pub state_default: Int,
 
     /// The power levels for specific users.
     ///
     /// This is a mapping from `user_id` to power level for that user.
+    ///
+    /// Must NOT contain creators of the room in room versions where the
+    /// `explicitly_privilege_room_creators` field of [`AuthorizationRules`] is set to `true`. This
+    /// would result in an error when trying to convert this to a [`RoomPowerLevelsEventContent`].
+    ///
+    /// When built from [`RoomPowerLevelsSource::None`]:
+    ///
+    /// * If `explicitly_privilege_room_creators` is set to `false` for the room version, defaults
+    ///   to setting the power level to `100` for the creator of the room.
```

Comment:
Should we also do this? Since `rules.additional_room_creators` is a separate field, it's theoretically possible for this to happen in a custom room version:
```suggestion
    ///   to setting the power level to `100` for the creator(s) of the room.
```

Reply:
Ah yes, good catch.